### DISCUSSION TITLE
[24 Gennaio 2023] Cache

### DISCUSSION THREAD

**Discussion by Elia-Belli** [sapienzastudentsnetwork/architettura-degli-elaboratori#57] [Cache (M-Z)]
![Esame](https://github.com/Elia-Belli/ae2223/assets/107944476/4cc3dccf-f66d-4a54-9334-3dfa0a246229)
---
### Soluzione del Prof
Le soluzioni fornite dal professore potrebbero essere incomplete o poco chiare, quindi sono benvenute soluzioni che le completano, oppure soluzioni alternative e/o con spiegazioni più approfondite: [Soluzione.pdf](https://github.com/sapienzastudentsnetwork/archelab2223/files/11665998/AE.2021-22.-.Appello.4.-.2023-01-24.-.Esercizio.cache.-.Scheda.di.calcolo.pdf)

**Comment by Frisbon** [comment]
Registrazione con svolgimento passo a passo allegata.
Letteralmente esercizio facile.it

https://youtu.be/AwOWSbpGIAA

![for you my love](https://github.com/sapienzastudentsnetwork/archelab2223/assets/84731341/24a86ef2-cdab-4a00-83de-258028e55032)

**Comment by FeddyLix17** [comment]
## 1) Accessi in memoria

dopo essersi ricavati le formule usando un approccio grafico

![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/f40d98ab-a28c-4713-a95e-283d8069d6bd)

si riempie la tabella ottenendo

![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/d6f64dd9-7aa6-4a35-ad9a-69772d882945)

## 2) Dimensioni delle cache

considerando che ogni linea della cache è formata

- dal bit di validità

- dal *Least Recently Used* bit (solo L1)

- dal tag

- dal blocco stesso

si ottiene, rispettivamente per le 2 cache
 
- $\Large \text{L1}$
  
   il numero delle linee è dato dal prodotto tra le vie e i set, da cui

$$
    \Large  (1 + 1 + 27 + 4 * 4 * 8) * (3 * 2) \implies (29 + 128) * 6 \implies  157 * 6 \implies 942\ \text{bit}
$$

- $\Large \text{L2}$

$$
    \Large  (1 + 22 + 32 * 4 * 8) * 8 \implies (23 + 1024) * 8 \implies  1047 * 8 \implies 8376\ \text{bit}
$$

## 3) Tempi della sequenza di accessi

### (a) Tempo totale sequenza di accessi

rielencandoli brevemente dalla foto superiore, si hanno

- 6 accessi in memoria, con tempo $\Large 200 * 6 = 1200\ \text{ns}$

- 3 accessi in L1, con tempo  $\Large 2 * 3 = 6\ \text{ns}$

- 3 accessi in L2, con tempo  $\Large 50 * 3 = 150\ \text{ns}$

$$
    \Large T_{tot} = 1200 + 6 + 150 = 1356\ \text{ns}
$$

### (b) Tempo medio sequenza di accessi

$$
    \Large T_{med} = \frac{\text{T.\ Totale}}{\text{Num.\ accessi}} = \frac{1356}{12} = 113.00\ \text{ns}
$$

### (c) Istruzioni svolte nel tempo medio

$$
    \Large \text{Istruzioni\ eseguite}  = \frac{\text{T.\ medio} *\ \text{Frequenza}}{\text{Clock\ per\ Istruction}} = \frac{113 * 8}{8} = 113.00
$$

## 4) Word offset indirizzo 2040 Cache L2

$$
    \Large \text{Word Offset} = (\text{Address} \gg \text{Byte Offset})\ \\%\ 2^{\text{bit Word Offset}}
$$

$$
    \Large  (2040 \gg 2)\ \\%\ 2^{5} \implies 510\ \\%\ 32 \implies 30
$$

## 5) Pagine Virtuali

ricavandosi la formula usando un approccio grafico

![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/27bbb35c-e103-49b5-aab5-832144bc18a5)

si riempie la tabella ottenendo

![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/664d1ad9-5902-4158-9631-6f4ab1e60d5c)
